common/common.go

```diff
@@ -645,3 +645,59 @@ func (c *Counter) IncrementAndGet() int64 {
 	}
 	return newID
 }
+
+// ElementProcessor defines the function signature for processing an individual element with its index.
+type ElementProcessor[E any] func(index int, element E) error
+
+// ProcessElementsByBatch takes a slice of elements and processes them in batches of `batchSize` concurrently.
+// For example, if batchSize = 10 and list has 100 elements, 10 goroutines will process 10 elements concurrently
+// in each batch. Each batch completes before the next batch begins.
+// `process` is a function called for each individual element with its index and value.
+func ProcessElementsByBatch[S ~[]E, E any](batchSize int, list S, process ElementProcessor[E]) (errs error) {
```

Comment:
This is more about throttling than it is batching.
I'll think today about naming / wording, but it wasn't until the third time round that I really clocked fully that this is a concurrent throughput throttler, not a batcher.

Reply:
I also just realised:
Is this implementation mechanics truly optimal for your use-case?
Normally you'd see things processed with a concurrency of 10.
Your implementation here may end up with a concurrency of 1 or 2 at then end of each batch, instead of maintaining 10 throughout
Maybe this is going to be self-evident in the code that calls this.

Reply:
Have you tested this as happening? Here's a better example like this one above:

```md
List of elements: [E1, E2, E3, ..., E100]

Batch size: 10

Step-by-step process:
1. Divide the list into batches of 10 elements each:
   Batch 1: [E1, E2, E3, ..., E10]
   Batch 2: [E11, E12, E13, ..., E20]
   ...
   Batch 10: [E91, E92, E93, ..., E100]

2. Process the batches sequentially:
   Batch 1 -> Batch 2 -> Batch 3 -> ... -> Batch 10
   (Each batch completes fully before moving to the next batch.)

3. Within each batch:
   - Spawn 10 goroutines (one for each element in the batch).
   - Each goroutine processes a single element by calling the `process` function with its index and value.
   
   Example for Batch 1:
       Goroutine 1 -> `process(0, E1)`
       Goroutine 2 -> `process(1, E2)`
       Goroutine 3 -> `process(2, E3)`
       ...
       Goroutine 10 -> `process(9, E10)`
```

So concurrency with this one scales linearly to how big the size is. `100 batch size` in this case will spawn `100 go routines` and process them all at once. Whereas the original I think (?) had an inverse relationship with the batch size, the higher the batch size the lower the concurrency. `100 batch size`  will process the entire list sequentially but a batch size of 1 would process instantly.

Reply:
Yeah, I did understand from the start.
I just don't know why you'd want to separate things into groups of 10 and then process all 10 before processing the next batch.
If 9 subscriptions complete in 0.2s each concurrently, but gateio decides to take 4 seconds on just one in that batch, because it decided now's a good time to cache invalidate, or whatever, then you're sitting there twiddling your thumbs.

I *can* imagine situations where this pattern is necessary, where there's some inter-relation that 10 must happen fully before the next 10. I just don't think that's the case for you here.

What would be the downside of a channel throttle that meant that you're *always* processing no more than 10 subscriptions?
That would make this a standard mini-worker and queue pattern.
I'm sure you know, but: _Fill a channel with slice, spawn N goros which read from the channel and exit when they don't get anything, waitgroup them and it's done._

Happy to throw up a PR into your branch if it helps.

But maybe I'm being really stupid and missing something obvious about why we'd need fully completed batches.

Reply:
I just remembered the main reasoning for Gateio snapshots specifically with 100 len books starts to interfere with gorilla websocket buffer and clips data, now that rate limiting subscription requests have been removed, so its been limited to 200 sub blocks which spawns 200 go routines. I will add that into the commentary 😆. 

> If 9 subscriptions complete in 0.2s each concurrently, but gateio decides to take 4 seconds on just one in that batch, because it decided now's a good time to cache invalidate, or whatever, then you're sitting there twiddling your thumbs.

Has this happened in testing? We can always mitigate it with a context deadline? But this seems like a bottom quintile issue only on startup and not really an issue with a long running program as the changes with subs are going to be limited to scaling up and down on exchange pair additions, removals and strategy requirements.  

Will check out your example now thanks.